3rdParty/libnzmq/src/org/zeromq/NativeZmqLoader.java

```diff
@@ -1,41 +1,63 @@
 package org.zeromq;
 
 import java.io.*;
-import java.net.URL;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Enumeration;
-import java.util.List;
-import java.util.jar.JarEntry;
-import java.util.jar.JarFile;
+import java.nio.file.Files;
+import java.nio.file.Path;
 
 /**
  * Loads (i.e. {@link System#load(String)} the native libraries that {@link ZMQ} binds to.
+ * The appropriate library is loaded based on the OS name.
  */
 public class NativeZmqLoader {
     public static final String NO_EMBEDDED_LIB_FLAG = "ZMQ_NO_EMBEDDED";
-    public static boolean LOADED_EMBEDDED_LIBRARY = false;
+    private static boolean LOADED_EMBEDDED_LIBRARY = false;
+
+    /** @return whether this class has already loaded the embedded library */
+    public static boolean isLoaded() {
+        return LOADED_EMBEDDED_LIBRARY;
+    }
 
     /**
      * Load native libs for ZMQ, unless:
-     *   (1) this class has already loaded it once successfully or 
+     *   (1) this class has already loaded it once successfully (i.e. {@link #isLoaded()} is true), or
      *   (2) {@link #NO_EMBEDDED_LIB_FLAG} is set
-     * 
+     *
+     * The following OSes are supported: Linux, Mac OS X, Windows.
      */
     public void load() {
         if(!LOADED_EMBEDDED_LIBRARY && System.getProperty(NO_EMBEDDED_LIB_FLAG) == null) {
-            LOADED_EMBEDDED_LIBRARY = loadNativeEmbedded("/native/linux/zmq/libjzmq.so")
-                && loadNativeEmbedded("/native/linux/zmq/libzmq.so.5");
+            try {
+                final Path libDir = Files.createTempDirectory("zmq_native");
+                libDir.toFile().deleteOnExit();
+                load(System.getProperty("os.name").toLowerCase(), libDir);
+            } catch (IOException ioe) {
+                ioe.printStackTrace();
```

Comment:
should we throw new RuntimeException in the catch?

Reply:
Yes we should ... will change